srml/support/procedural/src/storage/transfo.rs

```diff
@@ -0,0 +1,619 @@
+		opt_build = Some(build.as_ref().map(|b|&b.expr.content).map(|b|quote!( #b ))
+			.unwrap_or_else(||quote!( (|config: &GenesisConfig<#traitinstance>| config.#ident.clone()) )));
+		let fielddefault = default_value.inner.get(0).as_ref().map(|d|&d.expr).map(|d|
+			if is_option {
+				quote!( #d.unwrap_or_default() )
+			} else {
+				quote!( #d )
+			}).unwrap_or_else(|| quote!( Default::default() ));
+				config_field_default.push(quote!( #ident: #fielddefault, ));
+
+			} else {
+				opt_build = build.as_ref().map(|b|&b.expr.content).map(|b|quote!( #b ));
+			}
+
+			if let Some(builder) = opt_build {
+				is_trait_needed = true;
+				if is_simple {
+					builders.push(quote!{{
+						use #scrate::codec::Encode;
+						let v = (#builder)(&self);
+						r.insert(Self::hash(<#name<#traitinstance>>::key()).to_vec(), v.encode());
+					}});
+				} else {
+					builders.push(quote!{{
+						use #scrate::codec::Encode;
+						let data = (#builder)(&self);
+						for (k, v) in data.into_iter() {
+							r.insert(Self::hash(&<#name<#traitinstance>>::key_for(k)).to_vec(), v.encode());
+						}
+					}});
+				}
+			}
+
+		}
+
+		let mut has_scall = false;
+		let mut scall = quote!{ ( |_, _, _|{} ) };
+		let mut genesis_extrafields = Vec::new();
+		let mut genesis_extrafields_default = Vec::new();
+
+		// extra genesis
+		if let Some(eg) = extra_genesis {
+			for ex_content in eg.content.content.lines.inner.iter() {
+				match ex_content {
+					AddExtraGenesisLineEnum::AddExtraGenesisLine(AddExtraGenesisLine {
+						attrs,
+						extra_field,
+						extra_type,
+						default_value,
+						..
+					}) => {
+						if ext::has_parametric_type(&extra_type, traitinstance) {
+							is_trait_needed = true;
+							has_trait_field = true;
+						}
+						let extrafield = &extra_field.content;
+						genesis_extrafields.push(quote!{
+							#attrs pub #extrafield: #extra_type,
+						});
+						let extra_default = default_value.inner.get(0).map(|d|&d.expr).map(|e|quote!{ #e })
+							.unwrap_or_else(|| quote!( Default::default() ));
+						genesis_extrafields_default.push(quote!{
+							#extrafield: #extra_default,
+						});
+					},
+					AddExtraGenesisLineEnum::AddExtraGenesisBuild(DeclStorageBuild{ expr, .. }) => {
+						if has_scall { panic!( "Only one build expression allowed for extra genesis" ) }
+						let content = &expr.content;
+						scall = quote!( ( #content ) );
+						has_scall = true;
+					},
+				}
+			}
+		}
+
+		let is_extra_genesis_needed = has_scall
+			|| config_field.len() > 0
+			|| genesis_extrafields.len() > 0
+			|| builders.len() > 0;
+		if is_extra_genesis_needed {
+			let (fparam, sparam, ph_field, ph_default) = if is_trait_needed {
+				if has_trait_field {
+					// no phantom data required
+					(
+						quote!(<#traitinstance: #traittype>),
+						quote!(<#traitinstance>),
+						quote!(),
+						quote!(),
+					)
+				} else {
+					// need phantom data
+					(
```

Comment:
Why do we need a phantom field n this case?

Reply:
It is related to builder expression for type that are not in genesis, in this quote snippet : 
```rust
let v = (#builder)(&self);
r.insert(Self::hash(<#name<#traitinstance>>::key()).to_vec(), v.encode());
```
There is probably ways to avoid this but I cannot really say.
The one sample I checked was a decl_storage without a genesis declaration but with build expression in its type.

Reply:
There is a bunch of false positive :)  , I will fix that later. Yet found one : 
srml/system
```
decl_storage! {
	trait Store for Module<T: Trait> as System {

		pub AccountNonce get(account_nonce): map T::AccountId => T::Index;

		ExtrinsicCount: Option<u32>;
		pub BlockHash get(block_hash) build(|_| vec![(T::BlockNumber::zero(), [69u8; 32])]): map T::BlockNumber => T::Hash;
		ExtrinsicData get(extrinsic_data): map u32 => Vec<u8>;
		RandomSeed get(random_seed) build(|_| [0u8; 32]): T::Hash;
		/// The current block number being processed. Set by `execute_block`.
		Number get(block_number) build(|_| 1u64): T::BlockNumber;
		ParentHash get(parent_hash) build(|_| [69u8; 32]): T::Hash;
		ExtrinsicsRoot get(extrinsics_root): T::Hash;
		Digest get(digest): T::Digest;

		Events get(events): Vec<EventRecord<T::Event>>;
	}
	add_extra_genesis {
		config(changes_trie_config): Option<ChangesTrieConfiguration>;

		build(|storage: &mut primitives::StorageMap, _: &mut primitives::ChildrenStorageMap, config: &GenesisConfig<T>| {
			use codec::Encode;

			storage.insert(well_known_keys::EXTRINSIC_INDEX.to_vec(), 0u32.encode());

			if let Some(ref changes_trie_config) = config.changes_trie_config {
				storage.insert(
					well_known_keys::CHANGES_TRIE_CONFIG.to_vec(),
					changes_trie_config.encode());
			}
		});
	}
}
```

```
#[cfg(feature = "std")]
#[serde(rename_all = "camelCase")]
#[serde(deny_unknown_fields)]
pub struct GenesisConfig<T: Trait> {
    #[serde(skip)]
    pub _genesis_phantom_data: ::runtime_support::storage::generator::PhantomData<T>,
    pub changes_trie_config: Option<ChangesTrieConfiguration>,
}
```


```
#[cfg(feature = "std")]
impl<T: Trait> Default for GenesisConfig<T> {
    fn default() -> Self {
        GenesisConfig {
            _genesis_phantom_data: Default::default(),
            changes_trie_config: Default::default(),
        }
    }
}
#[cfg(feature = "std")]
impl<T: Trait> ::runtime_support::runtime_primitives::BuildStorage for GenesisConfig<T> {
    fn build_storage(
        self,
    ) -> ::std::result::Result<
        (
            ::runtime_support::runtime_primitives::StorageMap,
            ::runtime_support::runtime_primitives::ChildrenStorageMap,
        ),
        String,
    > {
        let mut r: ::runtime_support::runtime_primitives::StorageMap = Default::default();
        let mut c: ::runtime_support::runtime_primitives::ChildrenStorageMap = Default::default();
        {
            use runtime_support::codec::Encode;
            let data = (|_| <[_]>::into_vec(box [(T::BlockNumber::zero(), [69u8; 32])]))(&self);
            for (k, v) in data.into_iter() {
                r.insert(Self::hash(&<BlockHash<T>>::key_for(k)).to_vec(), v.encode());
            }
        }
        {
            use runtime_support::codec::Encode;
            let v = (|_| [0u8; 32])(&self);
            r.insert(Self::hash(<RandomSeed<T>>::key()).to_vec(), v.encode());
        }
        {
            use runtime_support::codec::Encode;
            let v = (|_| 1u64)(&self);
            r.insert(Self::hash(<Number<T>>::key()).to_vec(), v.encode());
        }
        {
            use runtime_support::codec::Encode;
            let v = (|_| [69u8; 32])(&self);
            r.insert(Self::hash(<ParentHash<T>>::key()).to_vec(), v.encode());
        }
        (|storage: &mut primitives::StorageMap,
          _: &mut primitives::ChildrenStorageMap,
          config: &GenesisConfig<T>| {
            use codec::Encode;
            storage.insert(well_known_keys::EXTRINSIC_INDEX.to_vec(), 0u32.encode());
            if let Some(ref changes_trie_config) = config.changes_trie_config {
                storage.insert(
                    well_known_keys::CHANGES_TRIE_CONFIG.to_vec(),
                    changes_trie_config.encode(),
                );
            }
        })(&mut r, &mut c, &self);
        Ok((r, c))
    }
}
```

Reply:
Ahh you mean that the build closure is taking `GenesisConfig<T>`?
Yeah, just remove the `T`.

Reply:
There is a bit more than that, in the build exps like : 
```
  {
            use runtime_support::codec::Encode;
            let v = (|_| 1u64)(&self);
            r.insert(Self::hash(<Number<T>>::key()).to_vec(), v.encode());
        }
```
Aka the `Number<T> ,RandomSeed<T>...`

Reply:
Then we just need to make the function generic.

Reply:
I do no think switching to generic function (especially in a trait) is really worth it (no dyn anymore and it will probably feel hacky: case where the type got no use)). Maybe it could be something for another pr.